File: public/java/090-k-closest-points-to-origin.java
/*
 * LeetCode Problem #973: K Closest Points to Origin
 * Link: https://leetcode.com/problems/k-closest-points-to-origin/
 * Difficulty: Medium
 */
import java.util.*;

class Solution {

    private int dist2(int[] p) { return p[0] * p[0] + p[1] * p[1]; }

    // APPROACH 1: Sort by Distance² | O(n log n) time | O(1) space
    // EXPLAIN: Sort all points by squared Euclidean distance and return the first k.
    public int[][] kClosest1(int[][] points, int k) {
        Arrays.sort(points, Comparator.comparingInt(this::dist2));
        return Arrays.copyOfRange(points, 0, k);
    }

    // APPROACH 2: Max Heap of Size K | O(n log k) time | O(k) space
    // EXPLAIN: Maintain a max-heap of size k; evict the farthest point when size exceeds k.
    public int[][] kClosest2(int[][] points, int k) {
        PriorityQueue<int[]> maxHeap = new PriorityQueue<>((a, b) -> dist2(b) - dist2(a));
        for (int[] p : points) {
            maxHeap.offer(p);
            if (maxHeap.size() > k) maxHeap.poll();
        }
        return maxHeap.toArray(new int[0][]);
    }

    // APPROACH 3: Quickselect | O(n) average time | O(log n) space
    // EXPLAIN: Partition in place until the first k elements are the k closest (unsorted).
    public int[][] kClosest3(int[][] points, int k) {
        quickselect(points, 0, points.length - 1, k);
        return Arrays.copyOfRange(points, 0, k);
    }

    private void quickselect(int[][] pts, int lo, int hi, int k) {
        if (lo >= hi) return;
        int p = partitionDist(pts, lo, hi);
        if (p == k) return;
        else if (p < k) quickselect(pts, p + 1, hi, k);
        else quickselect(pts, lo, p - 1, k);
    }

    private int partitionDist(int[][] pts, int lo, int hi) {
        int pivot = dist2(pts[hi]), i = lo;
        for (int j = lo; j < hi; j++) {
            if (dist2(pts[j]) <= pivot) {
                int[] tmp = pts[i]; pts[i] = pts[j]; pts[j] = tmp; i++;
            }
        }
        int[] tmp = pts[i]; pts[i] = pts[hi]; pts[hi] = tmp;
        return i;
    }

    // APPROACH 4: Partial Sort | O(n log k) time | O(k) space
    // EXPLAIN: Use a PriorityQueue as a min-heap to collect the k smallest by distance.
    public int[][] kClosest4(int[][] points, int k) {
        PriorityQueue<int[]> minHeap = new PriorityQueue<>(Comparator.comparingInt(this::dist2));
        for (int[] p : points) minHeap.offer(p);
        int[][] result = new int[k][];
        for (int i = 0; i < k; i++) result[i] = minHeap.poll();
        return result;
    }

    // APPROACH 5: Manual Quickselect | O(n) average time | O(log n) space
    // EXPLAIN: Explicit in-place quickselect with last-element pivot; same as Approach 3 but separated.
    public int[][] kClosest5(int[][] points, int k) {
        manualQS(points, 0, points.length - 1, k);
        return Arrays.copyOfRange(points, 0, k);
    }

    private void manualQS(int[][] pts, int lo, int hi, int k) {
        if (lo >= hi) return;
        int pivot = dist2(pts[hi]), i = lo;
        for (int j = lo; j < hi; j++) {
            if (dist2(pts[j]) <= pivot) {
                int[] tmp = pts[i]; pts[i] = pts[j]; pts[j] = tmp; i++;
            }
        }
        int[] tmp = pts[i]; pts[i] = pts[hi]; pts[hi] = tmp;
        if (i == k) return;
        else if (i < k) manualQS(pts, i + 1, hi, k);
        else manualQS(pts, lo, i - 1, k);
    }
}

// Made with Bob

File: public/python/090-k-closest-points-to-origin.py
# LeetCode Problem #973: K Closest Points to Origin
# Difficulty: Medium
# Link: https://leetcode.com/problems/k-closest-points-to-origin/

from typing import List
import heapq
import random

def dist2(p: List[int]) -> int:
    return p[0] * p[0] + p[1] * p[1]

# ─────────────────────────────────────────────
# APPROACH 1: Sort by Distance² | O(n log n) time | O(1) space
# EXPLAIN: Sort all points by squared Euclidean distance and return the first k.
# WHEN: Simplest approach; fine when n is small or code clarity matters most.

def solve_1(points: List[List[int]], k: int) -> List[List[int]]:
    points.sort(key=dist2)
    return points[:k]

# ─────────────────────────────────────────────
# APPROACH 2: Max Heap of Size K | O(n log k) time | O(k) space
# EXPLAIN: Maintain a max-heap of size k; evict the farthest point whenever the heap exceeds k.
# WHEN: Best when k << n; avoids sorting the full array.

def solve_2(points: List[List[int]], k: int) -> List[List[int]]:
    max_heap: list = []
    for i, p in enumerate(points):
        heapq.heappush(max_heap, (-dist2(p), i))
        if len(max_heap) > k:
            heapq.heappop(max_heap)
    return [points[i] for _, i in max_heap]

# ─────────────────────────────────────────────
# APPROACH 3: Quickselect (nth_element) | O(n) average time | O(1) space
# EXPLAIN: Partially partition the array so the first k elements are the k closest (unsorted).
# WHEN: Best average-case performance; equivalent to C++ nth_element.

def solve_3(points: List[List[int]], k: int) -> List[List[int]]:
    def partition(lo: int, hi: int) -> int:
        pivot_idx = random.randint(lo, hi)
        points[pivot_idx], points[hi] = points[hi], points[pivot_idx]
        pivot = dist2(points[hi])
        i = lo
        for j in range(lo, hi):
            if dist2(points[j]) <= pivot:
                points[i], points[j] = points[j], points[i]
                i += 1
        points[i], points[hi] = points[hi], points[i]
        return i

    def quickselect(lo: int, hi: int) -> None:
        if lo >= hi:
            return
        p = partition(lo, hi)
        if p == k:
            return
        elif p < k:
            quickselect(p + 1, hi)
        else:
            quickselect(lo, p - 1)

    quickselect(0, len(points) - 1)
    return points[:k]

# ─────────────────────────────────────────────
# APPROACH 4: Partial Sort | O(n log k) time | O(1) space
# EXPLAIN: Use heapq.nsmallest which internally uses a partial sort and returns the k smallest items.
# WHEN: Cleanest one-liner using the standard library; mirrors C++ partial_sort behaviour.

def solve_4(points: List[List[int]], k: int) -> List[List[int]]:
    return heapq.nsmallest(k, points, key=dist2)

# ─────────────────────────────────────────────
# APPROACH 5: Manual Quickselect | O(n) average time | O(log n) space
# EXPLAIN: Explicit in-place quickselect partitioning by squared distance.
# WHEN: When you want full control over the algorithm without library calls.

def solve_5(points: List[List[int]], k: int) -> List[List[int]]:
    def partition(lo: int, hi: int) -> int:
        pivot = dist2(points[hi])
        i = lo
        for j in range(lo, hi):
            if dist2(points[j]) <= pivot:
                points[i], points[j] = points[j], points[i]
                i += 1
        points[i], points[hi] = points[hi], points[i]
        return i

    def quickselect(lo: int, hi: int) -> None:
        if lo >= hi:
            return
        p = partition(lo, hi)
        if p == k:
            return
        elif p < k:
            quickselect(p + 1, hi)
        else:
            quickselect(lo, p - 1)

    quickselect(0, len(points) - 1)
    return points[:k]

# Made with Bob
